pkg/agent/controller/egress/ipassigner/ip_assigner_linux.go

```diff
@@ -34,153 +30,149 @@ import (
+	return dummy, nil
 }
 
-// UnassignEgressIP unassign an egressIP and delete the persistent files.
-func (a *ipAssigner) UnassignEgressIP(egressName string) error {
-	egressIP, has := a.isAssignedIP(egressName)
-	if !has {
-		return nil
-	}
-	egressSpecIP := net.ParseIP(egressIP)
-	addr := netlink.Addr{IPNet: &net.IPNet{IP: egressSpecIP, Mask: net.CIDRMask(32, 32)}}
-	ifaceName := a.egressLink.Attrs().Name
-	if err := netlink.AddrDel(a.egressLink, &addr); err != nil {
-		return fmt.Errorf("failed to delete ip %v from interface %s: %v", addr, ifaceName, err)
+// loadIPAddresses gets the IP addresses on the dummy device and caches them in memory.
+func (a *ipAssigner) loadIPAddresses() error {
+	addresses, err := netlink.AddrList(a.dummyDevice, netlink.FAMILY_ALL)
+	if err != nil {
+		return err
 	}
-	if err := a.removeAssignedIPFile(egressName); err != nil {
-		return fmt.Errorf("failed to remove Egress IP assign file, egressName: %s, egressIP: %s: %v", egressName, egressIP, err)
+	newAssignIPs := sets.NewString()
+	for _, address := range addresses {
+		newAssignIPs.Insert(address.IP.String())
 	}
-	klog.V(2).InfoS("Deleted ip", "ip", addr, "interface", ifaceName)
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	a.assignedIPs = newAssignIPs
 	return nil
 }
 
-// isAssignedIP check that if an IP address has been assigned with a specific egressName.
-func (a *ipAssigner) isAssignedIP(egressName string) (egressIP string, has bool) {
+func (a *ipAssigner) isAssignedIP(ip string) bool {
 	a.mutex.RLock()
 	defer a.mutex.RUnlock()
-	if ip, ok := a.assignedEgress[egressName]; ok {
-		return ip, true
-	}
-	return
+	return a.assignedIPs.Has(ip)
 }
 
-func ipSavedFile(dir, name string) string {
-	return filepath.Join(dir, name)
+func (a *ipAssigner) addAssignedIP(ip string) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	a.assignedIPs.Insert(ip)
 }
 
-func (a *ipAssigner) saveIPAssignFile(egressIP, egressName string) error {
-	var buffer bytes.Buffer
-	buffer.WriteString(egressIP)
-	f, err := os.Create(ipSavedFile(a.egressRunDir, egressName))
-	if err != nil {
-		return err
-	}
-	defer f.Close()
-	_, err = io.Copy(f, &buffer)
-	if err == nil {
-		a.mutex.Lock()
-		defer a.mutex.Unlock()
-		a.assignedEgress[egressName] = egressIP
+func (a *ipAssigner) deleteAssignedIP(ip string) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	a.assignedIPs.Delete(ip)
+}
+
+// AssignIP ensures the provided IP is assigned to the dummy device.
+func (a *ipAssigner) AssignIP(ip string) error {
+	parsedIP := net.ParseIP(ip)
+	if parsedIP == nil {
+		return fmt.Errorf("invalid IP %s", ip)
+	}
+	if a.isAssignedIP(ip) {
+		klog.V(2).InfoS("The IP is already assigned", "ip", ip)
 		return nil
 	}
-	return err
+
+	addr := netlink.Addr{IPNet: &net.IPNet{IP: parsedIP, Mask: net.CIDRMask(32, 32)}}
+	if err := netlink.AddrAdd(a.dummyDevice, &addr); err != nil {
+		return fmt.Errorf("failed to add IP %v to interface %s: %v", addr, a.dummyDevice.Attrs().Name, err)
+	}
+	a.addAssignedIP(ip)
```

Comment:
Good point! We don't allocate an IP to multiple Egresses but there is no simple mechanism to prevent user from doing it (validating webhook can only check ip conflict with current state in controller which processes event asynchronously), and technically specifying an IP to multiple Egresses can also work so what you pointed out is a valid case. I have updated the code to take care of this case.